[THS-66] Feat: Added metric registry

Euan-M-Campbell opened the pull request.

basaldella commented on src/scorebook/metrics/__init__.py:
```diff
@@ -11,8 +11,8 @@
 from scorebook.metrics.precision import Precision
 
-__all__ = ["MetricBase", "Precision", "Accuracy", "get_metrics"]
+__all__ = ["MetricBase", "Precision", "Accuracy", "metric_registry.py", "MetricRegistry"]
```
Why exposing the file and the class inside the file like this?

basaldella commented on src/scorebook/metrics/metric_base.py:
```diff
@@ -1,17 +1,17 @@
+    name: str
 
-    def __init__(self, name: str):
+    def __init__(self) -> None:
```
Why did you remove `name` and then check for the attr? Having the name in the constructor seems to be the best way to achieve this

basaldella reviewed (changes requested):
Good draft; we need to:
- Add tests for the registry in a dedicated class
- Make the example in the examples folder work 

I also noticed that we don't check if a metric already exists in the registry, let's throw a `ValueError` if someone adds a metric with the same name

basaldella commented on src/scorebook/metrics/metric_base.py:
```diff
@@ -1,32 +1,36 @@
+        output: Optional[Any] = None,
+        label: Optional[Any] = None,
+        evaluated_items: Optional[List[EvaluatedItem]] = None,
+    ) -> Any:
```
Let's improve the contract for what we return - we know it always has to be a list of per-item scores and a list of aggregated metric scores. Maybe `Tuple[List[Dict]], List[Dict]]` ? Not 100% sure

basaldella commented on src/scorebook/metrics/precision.py:
```diff
@@ -1,53 +1,57 @@
-        if len(predictions) != len(references):
-            raise ValueError("Predictions and references must have the same length")
+        # Calculate aggregate precision score
+        if evaluated_items:
```
Let's raise a NotImplementedError here to signal that it's not finished yet

basaldella reviewed (changes requested):
See inline comments + write tests for accuracy.py

basaldella commented on src/scorebook/evaluator.py:
```diff
@@ -77,61 +76,52 @@ def inference_fn(item):
-                    },
-                }
-        return results
+        return {
```
We should use `score_type` here and return aggregate vs per item correctly

Euan-M-Campbell commented:
Done:
- Removed results from remote repo, added to gitignore
- example output dir can be specified
- Metric aggregate scores can now have multiple values
- score type can now be used to specify returning aggregate scores and or item scores
- Fixed typo

basaldella commented on examples/basic_run.py:
```diff
@@ -64,27 +77,10 @@ def inference_function(eval_item: dict) -> Any:
-    print("\nItem Scores:")
-    for item_score in mmlu_pro_eval_results.item_scores:
-        print(f"\n{item_score}")
+    with open(output_dir / "output.json", "w") as output_file:
```
not os agnostic again, use `os.path.join` (also in the print message below)

basaldella commented on src/scorebook/evaluator.py:
```diff
@@ -116,9 +116,20 @@ def inference_fn(item):
     if return_type == "dict":
-        return {
-            eval_result.eval_dataset.name: eval_result.to_dict() for eval_result in eval_results
+        return_formats = {
```
The format is still not exactly what we want - we should actually have a list of dicts, and every dict should be flat (no nested dicts - see the tutorial).

basaldella reviewed (changes requested):
See inline comments; good work, needs minimal improvements and we can close

Euan-M-Campbell commented on examples/basic_run.py:
```diff
@@ -11,6 +13,17 @@
+    parser.add_argument(
+        "--output-dir",
+        type=str,
+        default=str(Path(__file__).parent / "results"),
```
The comment is outdated/wrong in this context. Your code is OS-agnostic because you're using pathlib.Path, which automatically handles the correct path separators on all platforms.

os.path.join is only needed when working with raw strings for paths (the old style). With Path, using / is overloaded to join paths in a cross-platform way.

Example:

Path(__file__).parent / "results"
is equivalent to:

os.path.join(os.path.dirname(__file__), "results")
but cleaner and OS-independent.

You don’t need to change anything.
If you want to be ultra-explicit, you could do:

default_output = Path(__file__).parent.joinpath("results")
but functionally, it’s the same.

Would you like me to give you a short, professional response you can post to explain why Path is already OS-agnostic?

You said:
are you sure?
ChatGPT said:
Yes, 100% sure.

basaldella reviewed (approved):
Well done! 🥳
